doc/developer-guides/trusty.rst

```diff
@@ -0,0 +1,186 @@
+    +-------------------------------------------+
+
+.. note::
+   Trusty OS is running in Secure World in the architecture above.
+
+Trusty specific Hypercalls
+**************************
+
+1. ``HC_INITIALIZE_TRUSTY``
+
+   * This Hypercall is used by the User OS bootloader (``UOS_Loader``) to request ACRN to initialize Trusty
+   * The Trusty memory region range, entry point must be specified
+   * Hypervisor needs to save current vCPU contexts (Normal World)
+#. ``HC_WORLD_SWITCH``
+
+   * Simulate ARM SMC (Secure Monitor Call, or SMC) instruction to do world switch
+   * Hypervisor needs to save current world vCPU contexts, and load next world vCPU contexts
+   * Update rdi, rsi, rdx, rbx to next world vCPU contexts
+
+APIs
+====
+
+.. code-block:: c
+
+   hcall_initialize_trusty(vm_t *vm)
+   hcall_world_switch(vm_t *vm)
+
+
+Trusty Boot flow
+****************
+
+By design, the User OS bootloader (``UOS_Loader``) will trigger the Trusty boot process. The complete bootflow is illustrated below.
+
+.. graphviz:: trusty/boot-flow.dot
+   :name: trusty-boot-flow
+   :align: center
+   :caption: Trusty Boot Flow
+
+Details
+
+1. UOS_Loader
+
+   a. Load and verify trusty image from virtual disk
+   #. Allocate runtime memory for trusty
+   #. Do ELF relocation of trusty image and get entry address
+   #. Call ``HC_INITIALIZE_TRUSTY`` with trusty memory base and entry address
+#. ACRN (``HC_INITIALIZE_TRUSTY``)
+
+   a. Save World context for Normal World
+   #. Init World context for Secure World(RIP, RSP, EPT, etc.)
+   #. Resume to Secure World
+#. Trusty
+
+   a. Booting up
+   #. Call ``HC_WORLD_SWITCH`` to switch back to Normal World if boot completed
+#. ACRN (``HC_WORLD_SWITCH``)
+
+   a. Save World context for the World which caused this ``vmexit`` (Secure World)
+   #. Restore World context for next World (Normal World (UOS_Loader))
+   #. Resume to next World (UOS_Loader)
+#. UOS_Loader
+
+   a. Continue to boot
+
+EPT Hierarchy
+*************
+
+As per the Trusty design, Trusty can access Normal World's memory, but Normal World cannot
+access Secure World's memory. Hence it means Secure World EPTP page table hierarchy
+must contain normal world GPA address space, while Trusty world's GPA address space
+must be removed from the Normal world EPTP page table hierarchy.
+
+Design
+======
+
+Put Secure World's GPA to very high position: 511G-512G. The PML4/PDPT for Trusty
+World are separated from Normal World. PD/PT for low memory (<511G) are shared in
+both Trusty World's EPT and Normal World's EPT. PD/PT for high memory (>=511G) are
+valid for Trusty World's EPT only.
+
+Benefit
+=======
+
+This design will benefit the EPT changes of Normal World. There are requirement to
+modify Normal World's EPT during runtime such as memory increasing, attribute
+change, etc. If such behavior happened, only PD and PT for Normal World need to
+be updated.
+
+::
+
+    ABSTRACT EPT hierarchy for 2 Worlds:
```

Comment:
Text drawing is OK for now, but we should change this to a real image (graphviz or a picture).

Reply:
Can you do something like this in `graphviz`?

I'll put this on my To-Do list